---
title: "Selection in 1000G EUR"
output:
  html_notebook:
    df_print: paged
    code_folding: hide
  html_document:
    df_print: paged
    code_folding: hide
---
# Selection in 1000G EUR

```{r}
# annotate peaks
peaks <- annotate_peaks(modern_all, modern_all_peaks, mathieson_peaks)

# count the number of SNPs in each group
num_gwas <- nrow(filter(peaks$data, type == "GWAS"))
num_control <- nrow(filter(peaks$data, type == "Control"))

# calculate the Bonferroni significance threshold
p.bonferroni <- 0.05 / num_gwas

# get the genome-wide and Bonferroni significant SNPs
snps_genomewide <- filter(peaks$data, p.value < p.genomewide)
snps_bonferroni <- filter(peaks$data, p.value < p.bonferroni)

# the columns to use for results tables
table_cols <- c("region", "peak", "rsid", "type", "chrom", "pos", "genes", "gwascat", "logLR", "p.value", "s")
```

Using a genome-wide significance threshold of `r formatC(p.genomewide, format = "e", digits = 0)`, we detect:

* `r nrow(filter(snps_genomewide, type=='GWAS'))` significant SNPs in the GWAS group (n=`r num_gwas`)
* `r nrow(filter(snps_genomewide, type=='Control'))` significant SNPs in the Control group (n=`r num_control`)

Using a Bonferroni corrected significance threshold of `r formatC(p.bonferroni, format = "e", digits = 2)`, we detect:

* `r nrow(filter(snps_bonferroni, type=='GWAS'))` significant SNPs in the GWAS group (n=`r num_gwas`)
* `r nrow(filter(snps_bonferroni, type=='Control'))` significant SNPs in the Control group (n=`r num_control`)

```{r fig1, fig.width = 12, fig.height = 6.75}
manhattan_plot(peaks$data, c("type"), p.genomewide, p.bonferroni, size_snps=FALSE)
```

**Figure XX.** Manhattan plot of the p-values from running CLUES on ARG containing all samples in FIN, GBR, and TSI from 1000G Phase 3, for (a) GWAS SNPs from the GWAS Catalog; and (b) Control SNPs, frequency paired with the GWAS SNPs.

```{r}
peaks$data %>%
  filter(peak %in% c("known", "novel")) %>%
  group_by(region) %>%
  slice(which.min(p.value)) %>%
  select(all_of(table_cols)) %>%
  arrange(p.value) %>%
  mutate(p.value = formatC(p.value, format = "e", digits = 2), logLR = round(logLR)) %>%
  kable(format = "html", caption = paste0("Table XX. ", length(peaks$all), " genome-wide significant peaks in aDNA time series (", length(peaks$novel), " novel and ", length(peaks$known), " previously characterised)")) %>%
  kable_styling(bootstrap_options = c("striped", "responsive"))
```

```{r}
peaks$data %>%
  filter(peak %in% c("absent")) %>%
  group_by(region) %>%
  slice(which.min(p.value)) %>%
  select(all_of(table_cols)) %>%
  arrange(p.value) %>%
  mutate(p.value = formatC(p.value, format = "e", digits = 2), logLR = round(logLR)) %>%
  kable(format = "html", caption = paste0("Table XX. ", length(peaks$absent), " previouly characterised peaks which were not significant in this analysis")) %>%
  kable_styling(bootstrap_options = c("striped", "responsive"))
```

Despite the lack of genome-wide significant SNPs, the GWAS group is significantly enriched for selection when compared to the Control group:
```{r}
paired <- pairs %>%
  inner_join(select(peaks$data, c(rsid, p.value)), by = c("gwas" = "rsid")) %>%
  inner_join(select(peaks$data, c(rsid, p.value)), by = c("neutral" = "rsid"), suffix = c(".gwas", ".neutral"))

wilcox <- wilcox.test(paired$p.value.gwas, paired$p.value.neutral, paired = TRUE, alternative = "less")
wilcox
```
