---
title: "Selection in aDNA with Ancestral Paintings"
output:
  html_notebook:
    df_print: paged
    code_folding: hide
  html_document:
    df_print: paged
    code_folding: hide
---
# Selection in aDNA with Ancestral Paintings

```{r}
# get all the ancient ancestry models
ancestral <- filter(clues, mode == "Ancient")

# annotate ancestry specific peaks
peaks_all <- annotate_peaks(filter(ancestral, ancestry == "ALL"), ancient_all_peaks, mathieson_peaks)
peaks_whg <- annotate_peaks(filter(ancestral, ancestry == "WHG"), ancient_whg_peaks, mathieson_peaks)
peaks_ehg <- annotate_peaks(filter(ancestral, ancestry == "EHG"), ancient_ehg_peaks, mathieson_peaks)
peaks_chg <- annotate_peaks(filter(ancestral, ancestry == "CHG"), ancient_chg_peaks, mathieson_peaks)
peaks_ana <- annotate_peaks(filter(ancestral, ancestry == "ANA"), ancient_ana_peaks, mathieson_peaks)

# concatenate all the annotated peaks data back together
peaks_joined <- bind_rows(
  peaks_all$data,
  peaks_whg$data,
  peaks_ehg$data,
  peaks_chg$data,
  peaks_ana$data,
) %>% arrange(as.character(chrom), start)

# merge all the peak regions (across all ancestries)
all_regions <- havester_to_bed(
  bind_rows(
    ancient_all_peaks,
    ancient_whg_peaks,
    ancient_ehg_peaks,
    ancient_chg_peaks,
    ancient_ana_peaks,
    # mathieson_peaks,
  ) %>%
    # add 50kb buffer either side
    separate(col = range, into = c("start", "end"), sep = "-", convert = TRUE) %>%
    mutate(range = paste0(start - 5e5, "-", end + 5e5))
)

pcadapt_regions <- havester_to_bed(
    pcadapt_peaks %>% 
    # add 50kb buffer either side
    separate(col = range, into = c("start", "end"), sep = "-", convert = TRUE) %>%
    mutate(range = paste0(start - 5e5, "-", end + 5e5))
)

# join the merged regions
peaks_joined <- bedr.join.region(peaks_joined$bed, all_regions, verbose = FALSE) %>%
  filter(V4 != ".") %>%
  mutate(merged_peaks = paste0(V4, ":", as.numeric(V5) + 5e5, "-", as.numeric(V6) - 5e5)) %>%
  select(index, merged_peaks) %>%
  unique() %>%
  left_join(peaks_joined, ., by = c("bed" = "index"))

ancestral <- peaks_joined %>%
  
  # drop the ancestry specific columns, or else the rows will not pivot correctly
  select(-c(region, peak)) %>%

  # pivot paintings from rows into columns, so we have one row per rsID
  pivot_wider(names_from = ancestry, names_glue = "{ancestry}_{.value}", values_from = c(logLR, p.value, s), names_sort = TRUE) %>%

  # calculate the minimum p.value across all ancestries
  mutate(min_p.value = select(., matches("p.value")) %>% purrr::pmap_dbl(min))

# sort the pivoted columns so they are grouped by their "ancestry" prefix
pivot_order <- map(levels(clues$ancestry), ~ names(ancestral)[grep(paste0(.x, "_"), names(ancestral))]) %>% unlist()
ancestral <- select(ancestral, all_of(c(setdiff(names(ancestral), pivot_order), pivot_order)))

write_tsv(ancestral, "supplementary_table_XX_CLUES-ancient.tsv")

# save the df to Google sheets
# gsheet <- gs4_create("CLUES_Ancestral_Paintings", sheets = ancestral)

# count the number of SNPs in each group
num_gwas <- nrow(filter(ancestral, type == "GWAS"))
num_control <- nrow(filter(ancestral, type == "Control"))

# calculate the Bonferroni significance threshold
p.bonferroni <- 0.05 / num_gwas

# get the genome-wide and Bonferroni significant SNPs
snps_genomewide <- filter(ancestral, min_p.value < p.genomewide)
snps_bonferroni <- filter(ancestral, min_p.value < p.bonferroni)

# the columns to use for results tables
table_cols <- c("merged_peaks", "rsid", "type", "chrom", "pos", "genes", "gwascat", pivot_order, "min_p.value")
```

Using a genome-wide significance threshold of `r formatC(p.genomewide, format = "e", digits = 0)`, we detect:

* `r nrow(filter(snps_genomewide, type=='GWAS'))` significant SNPs in at least one ancestry the GWAS group (n=`r num_gwas`)
* `r nrow(filter(snps_genomewide, type=='Control'))` significant SNPs in at least one ancestry in the Control group (n=`r num_control`)

Using a Bonferroni corrected significance threshold of `r formatC(p.bonferroni, format = "e", digits = 2)`, we detect:

* `r nrow(filter(snps_bonferroni, type=='GWAS'))` significant SNPs in at least one ancestry in the GWAS group (n=`r num_gwas`)
* `r nrow(filter(snps_bonferroni, type=='Control'))` significant SNPs in at least one ancestry in the Control group (n=`r num_control`)

```{r fig3, fig.width = 12, fig.height = 12}
manhattan_plot(filter(peaks_joined, type == "GWAS"), c("ancestry", "type"), p.genomewide, p.bonferroni, colors = ancestry_colors, size_snps=FALSE, wrap=TRUE)
```

**Figure XX.** Manhattan plot of the p-values from running CLUES on an aDNA time series conditioned on ancestry paintings from all West Eurasian samples in the imputed dataset for GWAS SNPs from the GWAS Catalog.

```{r fig4, fig.width = 12, fig.height = 12}
manhattan_plot(filter(peaks_joined, type == "Control"), c("ancestry", "type"), p.genomewide, p.bonferroni, colors = ancestry_colors, size_snps=FALSE, wrap=1)
```

**Figure XX.** Manhattan plot of the p-values from running CLUES on an aDNA time series conditioned on ancestry paintings from all West Eurasian samples in the imputed dataset for Control SNPs, frequency paired with the GWAS SNPs. Dark green dots represent novel selection peaks (with respect to Mathieson et al. 2015), light green are known selection peaks, and orange are previously reported peaks which are not found to be significant in this analysis.

```{r}
# construct the header groups
header_groups <- c(" " = 7, setNames(rep(3, length(levels(clues$ancestry))), levels(clues$ancestry)), "min()" = 1)

# strip the prefixes from the column names
short_names <- str_replace(table_cols, "^\\w{3}_", "")

peaks_table <- ancestral %>%
  filter(!is.na(merged_peaks)) %>%
  group_by(merged_peaks) %>%
  slice(which.min(min_p.value)) %>%
  ungroup() %>%
  select(all_of(table_cols)) %>%
  arrange(chrom, pos) %>%
  mutate_if(is.numeric, list(~ ifelse(. < 1, formatC(., format = "e", digits = 2), .)))

peaks_table %>%
  kable(format = "html", col.names = short_names, caption = paste0("Table XX. ", length(all_regions), " genome-wide significant peaks in aDNA time series conditioned on ancestry paintings")) %>%
  kable_styling(bootstrap_options = c("striped", "responsive")) %>%
  add_header_above(header_groups)
```
```{r}

for (row in 1:nrow(peaks_table)) {
  sig.yes <- c()
  sig.no  <- c()
  
  for (anc in c("ALL", "WHG", "EHG", "CHG", "ANA")) {
    if (peaks_table[row, paste0(anc, "_p.value")] < p.genomewide) {
      sig.yes <- c(sig.yes, anc)
    } else {
      sig.no <- c(sig.no, anc)
    }
  }
  
  # if (length(sig.no) == 0) {
  #   anc_regions <- "Significant peaks detected in all ancestries."
  # } else if (length(sig.yes) == 1) {
  #   anc_regions <- paste0("Significant peaks detected only one ancestry (", sig.yes, ")")
  # } else {
  #   anc_regions <- paste0("Significant peaks detected in ", length(sig.yes), " ancestries only (", paste(sig.yes, collapse=", "), ").")
  # }
  
  print(paste0("Figure XX. Zoomed in Manhattan plot of the p-values from running CLUES on an aDNA time series conditioned on ancestry paintings from all West Eurasian samples in the imputed dataset for GWAS SNPs in the peak region ", peaks_table[row, "merged_peaks"],"."))
  
  print(paste0("Selection peak ", row, " spans the region ", peaks_table[row, "merged_peaks"],
               ", with the most significant SNP being ", peaks_table[row, "rsid"],
               " (", peaks_table[row, "genes"],
               "; min(p)=", peaks_table[row, "min_p.value"],
               "; s=XXX), associated with ", peaks_table[row, "gwascat"], "."))
}
```


```{r fig5, fig.width = 12, fig.height = 4.4}
# {r fig5, fig.width = 12, fig.height = 105}
# manhattan_plot(peaks_joined %>% drop_na(merged_peaks), c("merged_peaks", "ancestry"), p.genomewide, p.bonferroni, num_label = 3, wrap = 5)

for (focal_peak in str_sort(unique(peaks_joined$merged_peaks), numeric = TRUE, na_last = NA)) {
  plt <- manhattan_plot(filter(peaks_joined, merged_peaks==focal_peak), c("merged_peaks", "ancestry"), p.genomewide, p.bonferroni, num_label = 3, wrap = 5)
  print(plt)
}
```


Similarly to the 1000G EUR analysis, the aDNA time series with ancestral painting GWAS group is significantly enriched for selection when compared to the Control group:
```{r}
paired <- pairs %>%
  inner_join(select(ancestral, c(rsid, min_p.value)), by = c("gwas" = "rsid")) %>%
  inner_join(select(ancestral, c(rsid, min_p.value)), by = c("neutral" = "rsid"), suffix = c(".gwas", ".neutral"))

wilcox <- wilcox.test(paired$min_p.value.gwas, paired$min_p.value.neutral, paired = TRUE, alternative = "less")
wilcox
```
